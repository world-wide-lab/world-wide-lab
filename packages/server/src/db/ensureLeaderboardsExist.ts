import type { Sequelize } from "sequelize";

async function ensureLeaderboardsExist(
  sequelize: Sequelize,
  leaderboardIds: string[],
) {
  await Promise.all(
    leaderboardIds.map((leaderboardId) =>
      ensureLeaderboardExists(sequelize, leaderboardId),
    ),
  );
}

async function ensureLeaderboardExists(
  sequelize: Sequelize,
  leaderboardId: string,
) {
  const existingLeaderboardCount = (
    await sequelize.models.Leaderboard.findAndCountAll({
      where: {
        leaderboardId,
      },
    })
  ).count;

  if (existingLeaderboardCount > 0) {
    // Leaderboard already exists
    return;
  }

  // Create leaderboard
  await sequelize.models.Leaderboard.create({
    leaderboardId,
  });
}

export { ensureLeaderboardsExist, ensureLeaderboardExists };
