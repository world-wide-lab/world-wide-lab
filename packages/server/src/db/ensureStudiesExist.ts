import type { Sequelize } from "sequelize";

async function ensureStudiesExist(sequelize: Sequelize, studyIds: string[]) {
  await Promise.all(
    studyIds.map((studyId) => ensureStudyExists(sequelize, studyId)),
  );
}

async function ensureStudyExists(sequelize: Sequelize, studyId: string) {
  const existingStudyCount = (
    await sequelize.models.Study.findAndCountAll({
      where: {
        studyId,
      },
    })
  ).count;

  if (existingStudyCount > 0) {
    // Study already exists
    return;
  }

  // Create study
  await sequelize.models.Study.create({
    studyId,
  });
}

export { ensureStudiesExist, ensureStudyExists };
