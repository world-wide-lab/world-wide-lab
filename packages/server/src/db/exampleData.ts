import { Sequelize } from "sequelize";
import { logger } from "../logger";

async function generateExampleData(
  sequelize: Sequelize,
  studyId: string = "example",
) {
  const exampleStudyCount = (
    await sequelize.models.Study.findAndCountAll({
      where: {
        studyId,
      },
    })
  ).count;

  if (exampleStudyCount > 0) {
    // Data has already been generated
    return;
  }
  logger.info("Generating Example Data");
  await sequelize.models.Study.create({
    studyId,
  });

  async function generateParticipantData(data: {
    privateInfo: Object;
    sessions: Array<number>;
    createdAt?: Date;
    finished_after_sessions?: number;
  }) {
    const { privateInfo, sessions, createdAt, finished_after_sessions } = data;
    const participant = await sequelize.models.Participant.create({
      privateInfo,
      createdAt,
    });
    for (let sessionIndex = 0; sessionIndex < sessions.length; sessionIndex++) {
      const n_responses = sessions[sessionIndex];
      const session = await sequelize.models.Session.create({
        // @ts-ignore
        participantId: participant.participantId,
        studyId,
        createdAt,
        finished:
          finished_after_sessions === undefined
            ? false
            : n_responses >= finished_after_sessions,
      });
      const responses = [];
      for (let index = 1; index <= n_responses; index++) {
        responses.push({
          // @ts-ignore
          sessionId: session.sessionId,
          name: `trial-${index}`,
          payload: {
            response: `Response #${index}`,
          },
        });
      }
      await sequelize.models.Response.bulkCreate(responses);
    }
  }

  function daysAgo(daysAgo: number): Date {
    const date = new Date();
    date.setDate(date.getDate() - daysAgo);
    return date;
  }

  await generateParticipantData({
    privateInfo: {
      description: "Example User #1",
    },
    sessions: [3],
    finished_after_sessions: 3,
    createdAt: daysAgo(4),
  });
  await generateParticipantData({
    privateInfo: {
      description: "Example User #2",
    },
    sessions: [3],
    finished_after_sessions: 3,
    createdAt: daysAgo(3),
  });
  await generateParticipantData({
    privateInfo: {
      description: "Example User #3",
    },
    sessions: [3, 3, 1, 3],
    finished_after_sessions: 3,
    createdAt: daysAgo(2),
  });
  await generateParticipantData({
    privateInfo: {
      description: "Example User #4",
    },
    sessions: [3, 1],
    finished_after_sessions: 3,
    createdAt: daysAgo(1),
  });
}

export default generateExampleData;
