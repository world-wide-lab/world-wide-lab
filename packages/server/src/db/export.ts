import type { Response } from "express";
import { json2csv } from "json-2-csv";
import { QueryTypes, Sequelize } from "sequelize";
import config from "../config";

// Export data from the database in chunks
async function paginatedExport(
  res: Response,
  queryData: (offset: number, limit: number) => Promise<object[]>,
  format: "json" | "csv",
  limit: number = Infinity,
  initialOffset = 0,
) {
  const onStart = () => {
    if (format === "json") {
      res.status(200).contentType("application/json");

      res.write("[");
    } else if (format === "csv") {
      res.status(200).contentType("text/csv");
    }
  };

  // Process & return each chunk of data
  const onData = async (data: object[], offset: number) => {
    const isFirstChunk = offset === initialOffset;

    if (format === "json") {
      // Convert data to JSON string
      const json = JSON.stringify(data);

      if (!isFirstChunk) {
        // Add comma in between JSON chunks
        res.write(",");
      }

      // Remove first and last characters from JSON string
      // They should be "[" and "]" respectively.
      // Since we want to combine data from multiple chunks, we have to
      // remove them within every chunk and only add them once in
      // onStart and onEnd.
      res.write(json.substring(1, json.length - 1));
    } else if (format === "csv") {
      if (!isFirstChunk) {
        // Add newline in between CSV chunks
        res.write("\n");
      }

      // Convert to CSV
      res.write(
        json2csv(data, {
          prependHeader: isFirstChunk,
          expandNestedObjects: false,
          useDateIso8601Format: true,
        }),
      );
    } else {
      throw new Error(`Unknown format: ${format}`);
    }
  };
  // Complete the data export
  const onEnd = () => {
    if (format === "json") {
      res.write("]");
    }

    // Mark the response as finished
    res.end();
  };

  return await chunkedQuery({
    queryData,
    onData,
    onStart,
    onEnd,
    limit,
    initialOffset,
  });
}

async function chunkedQuery({
  queryData,
  onData,
  onStart = () => {},
  onEnd = () => {},
  limit = Infinity,
  initialOffset = 0,
}: {
  queryData: (offset: number, limit: number) => Promise<object[]>;
  onData: (data: object[], offset: number) => Promise<void>;
  onStart?: () => void;
  onEnd?: () => void;
  limit?: number;
  initialOffset?: number;
}) {
  let pageSize = config.database.chunkSize;
  let offset = initialOffset;
  const absoluteLimit = initialOffset + limit;

  let nRowsResult: number;
  let isFirstIteration = true;
  if (limit < pageSize) {
    pageSize = limit;
  }

  do {
    // If we'd overshoot with the next page, reduce the page size to exactly fit the limit
    if (offset + pageSize > absoluteLimit) {
      pageSize = absoluteLimit - offset;
    }

    // Retrieve the data
    const data = await queryData(offset, pageSize);
    if (!Array.isArray(data)) {
      throw new Error("Data is always expected to be returned as an Array");
    }
    nRowsResult = data.length;

    if (isFirstIteration) {
      // Call onStart after we received data for the first time to still allow
      // sending error status codes if query code fails.
      onStart();
      isFirstIteration = false;
    }

    // Do something with the data (usually returning it to the user)
    if (nRowsResult > 0) {
      await onData(data, offset);
    }

    // Increase the offset in case we will continue
    offset += pageSize;

    // Check whether we already got all data
    // e.g. we either got an empty result or our result was less than the limit
  } while (
    nRowsResult > 0 &&
    nRowsResult === pageSize &&
    offset < absoluteLimit
  );

  onEnd();
}

async function generateExtractedPayloadQuery(
  sequelize: Sequelize,
  studyId: string,
): Promise<string> {
  // Get all keys which are present in the payloads of the responses
  const keysResult = await sequelize.query(
    `
      SELECT DISTINCT
        payload_json.key as key
      FROM
        wwl_responses
          INNER JOIN wwl_sessions ON (wwl_sessions."sessionId" = wwl_responses."sessionId"),
        json_each(payload) payload_json
      WHERE wwl_sessions."studyId" = :studyId
      ORDER BY key ASC;
    `,
    {
      type: QueryTypes.SELECT,
      replacements: { studyId },
    },
  );

  // Create a the list of keys in the payload in a safe format
  const jsonKeys = keysResult
    .map((row) => ("key" in row ? row.key : undefined))
    .filter((key) => key !== undefined);
  const jsonFieldsString = jsonKeys
    .map((jsonKey) => `wwl_responses."payload"->>'${jsonKey}' AS "${jsonKey}"`)
    .join(", ");

  // Collect list of table fields, so we can select them without the payload
  const modelAttributes = sequelize.models.Response.getAttributes();
  const tableFields = Object.keys(modelAttributes);
  const fields = tableFields
    .filter((field) => field !== "payload")
    .map((field) => `wwl_responses."${field}"`);
  const tableFieldsString = fields.join(", ");

  // Combine the table and json fields
  const fieldsString =
    jsonKeys.length > 0
      ? `${tableFieldsString}, ${jsonFieldsString}`
      : `${tableFieldsString}`;

  // Create the final query
  return `
    SELECT
      ${fieldsString}
    FROM wwl_responses
      INNER JOIN wwl_sessions ON (wwl_sessions."sessionId" = wwl_responses."sessionId")
    WHERE wwl_sessions."studyId" = :studyId
  `;
}

export { paginatedExport, generateExtractedPayloadQuery };
