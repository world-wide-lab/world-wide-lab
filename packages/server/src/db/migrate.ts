import pathLib from "node:path";
import { SequelizeStorage, Umzug } from "umzug";

import process from "node:process";
import { fileURLToPath } from "node:url";
import { logger } from "../logger.js";
import { getDirectory } from "../util.js";
import sequelize from "./index.js";

const dirname = getDirectory(import.meta.url);

const dialect = sequelize.getDialect();

const umzug = new Umzug({
  // Support both common and dialect-specific migrations
  migrations: {
    glob: [`migrations/*.@(common|${dialect}).@(js|ts)`, { cwd: dirname }],
    resolve: ({ name, path, context }) => {
      const getMigration = async () => import(path as string);
      const nameWithoutExtension = pathLib.parse(path as string).name;

      // adjust the parameters Umzug will
      // pass to migration methods when called
      return {
        // Remove extension of migration names (to avoid double applying between js and ts)
        name: nameWithoutExtension,
        up: async () => {
          const migration = await getMigration();
          await migration.up({ context });
        },
        down: async () => {
          const migration = await getMigration();
          await migration.down({ context });
        },
      };
    },
  },
  context: sequelize.getQueryInterface(),
  storage: new SequelizeStorage({
    sequelize,
    modelName: "InternalMigrations",
    tableName: "wwl_internal_migrations",
  }),
  // Map winston onto the umzug internal logger version
  logger: {
    info: (message) => logger.info,
    warn: (message) => logger.warn,
    error: (message) => logger.error,
    debug: (message) => logger.info,
  },
});

async function up() {
  const pending = await umzug.pending();
  if (pending.length > 0) {
    const migrationNames: Array<string> = Object.values(pending).map(
      (migration) => migration.name,
    );
    logger.info("Applying pending migrations: ", migrationNames);
  } else {
    logger.info("No pending migrations");
  }

  // Checks migrations and run them if they are not already applied. To keep
  // track of the executed migrations, a table (and sequelize model) called SequelizeMeta
  // will be automatically created (if it doesn't exist already) and parsed.
  await umzug.up();
}

let latestMigration: string;
async function getLatestMigration(includeSuffix: boolean): Promise<string> {
  if (latestMigration === undefined) {
    const executed = await umzug.executed();
    latestMigration = executed[executed.length - 1].name;
  }
  if (includeSuffix) {
    return latestMigration;
  }
  return latestMigration.substring(0, latestMigration.lastIndexOf("."));
}

// Detect whether the file is called directly
// TODO: Find a cleaner solution for this and determine whether we even need this
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  // Can be called via e.g.
  // node dist/db/migrate.js create --name migration-name.ts --skip-verify
  // after npm run build, or the following before building
  // npx node --loader ts-node/esm src/db/migrate.ts create --name migration-name.ts --skip-verify
  umzug.runAsCLI();
}

// export the type helper exposed by umzug, which will have the `context` argument typed correctly
export type Migration = typeof umzug._types.migration;

export { umzug, up, getLatestMigration };
