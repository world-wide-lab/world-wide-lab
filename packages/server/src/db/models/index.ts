import { Sequelize, DataTypes } from "sequelize";

const columnComments = {
  studyId: `The unique identifier for each study. This id is used to link sessions with studies. Must be unique across all studies.`,
  participantId: `The unique identifier for each participant. This id is used to link sessions with participants. Generated automatically.`,
  sessionId: `The unique identifier for each session. This id is used to identify responses. Generated automatically.`,
  responseId: `The unique identifier for each response. Generated automatically.`,

  createdAt: `The timestamp this record has been created. Generated automatically.`,
  updatedAt: `The timestamp this record has last been updated or changed. Generated automatically.`,
  privateInfo: `Additional information for this record, stored as a JSON object.`,
  publicInfo: `Additional public information for this record, stored as a JSON object. This field must not contain sensitive information as its contents can be queried from the public API.`,
  deletionProtection: `Should the study be protected from deletion? If this is set to true, the study cannot be deleted from the admin interface until this is turned off again. This is useful to prevent accidental deletion of studies that have already been published.`,
};

function defineModels(sequelize: Sequelize) {
  const Study = sequelize.define(
    "Study",
    {
      studyId: {
        primaryKey: true,
        type: DataTypes.STRING,
        validate: {
          is: /^[a-zA-Z0-9-_]+$/,
        },
        unique: true,
        allowNull: false,
        defaultValue: null,
        comment: columnComments.studyId,
      },
      createdAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        comment: columnComments.createdAt,
      },
      updatedAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        onUpdate: "CASCADE",
        comment: columnComments.updatedAt,
      },
      privateInfo: {
        type: DataTypes.JSON,
        allowNull: true,
        comment: columnComments.privateInfo,
      },
      publicInfo: {
        type: DataTypes.JSON,
        allowNull: true,
        comment: columnComments.publicInfo,
      },
      deletionProtection: {
        type: DataTypes.BOOLEAN,
        allowNull: false,
        defaultValue: true,
        comment: columnComments.deletionProtection,
      },
    },
    {
      tableName: "wwl_studies",
    },
  );

  const Participant = sequelize.define(
    "Participant",
    {
      participantId: {
        type: DataTypes.UUID,
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4,
        comment: columnComments.participantId,
      },
      createdAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        comment: columnComments.createdAt,
      },
      updatedAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        onUpdate: "CASCADE",
        comment: columnComments.updatedAt,
      },
      privateInfo: {
        type: DataTypes.JSON,
        allowNull: true,
        comment: columnComments.privateInfo,
      },
      publicInfo: {
        type: DataTypes.JSON,
        allowNull: true,
        comment: columnComments.publicInfo,
      },
    },
    {
      tableName: "wwl_participants",
    },
  );

  const Session = sequelize.define(
    "Session",
    {
      sessionId: {
        type: DataTypes.UUID,
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4,
        comment: columnComments.sessionId,
      },
      createdAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        comment: columnComments.createdAt,
      },
      updatedAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        onUpdate: "CASCADE",
        comment: columnComments.updatedAt,
      },
      privateInfo: {
        type: DataTypes.JSON,
        allowNull: true,
        comment: columnComments.privateInfo,
      },
      publicInfo: {
        type: DataTypes.JSON,
        allowNull: true,
        comment: columnComments.publicInfo,
      },
      finished: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
        comment: `Has this session has been finished? Note, that this field only gets updated when the /session/finish API endpoint is called.`,
      },
      participantId: {
        type: DataTypes.UUID,
        allowNull: true,
        comment: columnComments.participantId,
      },
      studyId: {
        type: DataTypes.STRING,
        allowNull: false,
        comment: columnComments.studyId,
      },
    },
    {
      tableName: "wwl_sessions",
    },
  );

  const Response = sequelize.define(
    "Response",
    {
      responseId: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true,
        allowNull: false,
        comment: columnComments.responseId,
      },
      createdAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        comment: columnComments.createdAt,
      },
      updatedAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        onUpdate: "CASCADE",
        comment: columnComments.updatedAt,
      },
      name: {
        type: DataTypes.STRING,
        allowNull: true,
      },
      payload: {
        type: DataTypes.JSON,
        allowNull: true,
      },
      sessionId: {
        type: DataTypes.UUID,
        allowNull: false,
        comment: columnComments.sessionId,
      },
    },
    {
      tableName: "wwl_responses",
    },
  );

  // Associations
  Participant.hasMany(Session, { foreignKey: "participantId" });
  Session.belongsTo(Participant, { foreignKey: "participantId" });

  Study.hasMany(Session, { sourceKey: "studyId", foreignKey: "studyId" });
  Session.belongsTo(Study, { targetKey: "studyId", foreignKey: "studyId" });

  Session.hasMany(Response, { foreignKey: "sessionId" });
  Response.belongsTo(Session, { foreignKey: "sessionId" });

  const InternalAdminSession = sequelize.define(
    "InternalAdminSession",
    {
      sid: {
        type: DataTypes.STRING(36),
        primaryKey: true,
      },
      expires: DataTypes.DATE,
      data: DataTypes.TEXT,
      createdAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
      },
      updatedAt: {
        type: DataTypes.DATE,
        defaultValue: DataTypes.NOW,
        onUpdate: "CASCADE",
      },
    },
    {
      tableName: "wwl_internal_admin_sessions",
    },
  );
}

export { defineModels, columnComments };
