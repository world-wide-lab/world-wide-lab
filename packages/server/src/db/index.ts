import { Sequelize, Options as SequelizeOptions } from "sequelize";

import { defineModels } from "./models";
import config from "../config";
import { log_sql } from "../logger";

const url = config.database.url;
const options: SequelizeOptions = {
  logging: log_sql,
};

// There is a bug in handling filepaths with spaces when using sqlite via
// a URL string. Therefore we manually pass it the database URL for sqlite.
const isSqlite = url.startsWith("sqlite:");
if (isSqlite) {
  // Manually parse URL instead and add it to options
  options.dialect = "sqlite";
  options.storage = url.replace("sqlite://", "").replace("sqlite:", "");
}
const sequelize = isSqlite
  ? new Sequelize(options)
  : new Sequelize(url, options);

defineModels(sequelize);

export default sequelize;
