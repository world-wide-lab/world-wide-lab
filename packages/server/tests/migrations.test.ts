// Set up fake environment variables
import "./setup_env";

import sequelize from "../src/db";
import { up, umzug } from "../src/db/migrate";

describe("Database Migrations", () => {
  it("should authenticate to database", async () => {
    await sequelize.authenticate();
  });

  it("should have migrations to apply", async () => {
    const pending = await umzug.pending();
    expect(pending.length).toBeGreaterThan(0);
  });

  it("should session migrations without issues", async () => {
    await up();
  });

  it("should always create the same tables", async () => {
    const tableNames: Array<string> = await sequelize
      .getQueryInterface()
      .showAllTables();
    tableNames.sort();
    expect(tableNames).toMatchSnapshot();
  });

  it("should always create the same table structures", async () => {
    const tableNames: Array<string> = await sequelize
      .getQueryInterface()
      .showAllTables();
    tableNames.sort();

    const tableInfos = await Promise.all(
      tableNames.map(async (tableName) => {
        const tableInfo = await sequelize
          .getQueryInterface()
          .describeTable(tableName);
        return tableInfo;
      }),
    );
    expect(tableInfos).toMatchSnapshot();
  });

  it("should have tables for all models", async () => {
    expect(await sequelize.models.Study.count()).toBe(0);
    expect(await sequelize.models.Participant.count()).toBe(0);
    expect(await sequelize.models.Session.count()).toBe(0);
    expect(await sequelize.models.Response.count()).toBe(0);
  });

  it("should be in-sync with the models afterwards", async () => {
    // sequelize.sync with alter: false shouldn't fail
    await sequelize.sync({ alter: false });
  });
});
