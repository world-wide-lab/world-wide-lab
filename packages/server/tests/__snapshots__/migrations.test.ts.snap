// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Database Migrations > should always create the same table structures 1`] = `
[
  {
    "createdAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "deploymentConfig": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "deploymentId": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": true,
      "type": "INTEGER",
      "unique": false,
    },
    "name": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "privateInfo": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "stackConfig": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "status": {
      "allowNull": false,
      "defaultValue": "undeployed",
      "primaryKey": false,
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "type": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "updatedAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
  },
  {
    "createdAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "data": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "TEXT",
      "unique": false,
    },
    "expires": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "sid": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": true,
      "type": "VARCHAR(36)",
      "unique": true,
    },
    "updatedAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
  },
  {
    "name": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": true,
      "type": "VARCHAR(255)",
      "unique": true,
    },
  },
  {
    "createdAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "leaderboardId": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "leaderboardScoreId": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": true,
      "type": "INTEGER",
      "unique": false,
    },
    "publicGroupName": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "publicIndividualName": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "score": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "INTEGER",
      "unique": false,
    },
    "sessionId": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "UUID",
      "unique": false,
    },
    "updatedAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
  },
  {
    "createdAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "leaderboardId": {
      "allowNull": false,
      "defaultValue": null,
      "primaryKey": true,
      "type": "VARCHAR(255)",
      "unique": true,
    },
    "privateInfo": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "studyId": {
      "allowNull": true,
      "defaultValue": null,
      "primaryKey": false,
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "updatedAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
  },
  {
    "createdAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "participantId": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": true,
      "type": "UUID",
      "unique": true,
    },
    "privateInfo": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "publicInfo": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "updatedAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
  },
  {
    "createdAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "name": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "payload": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "responseId": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": true,
      "type": "INTEGER",
      "unique": false,
    },
    "sessionId": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": false,
      "references": {
        "key": "sessionId",
        "model": "wwl_sessions",
      },
      "type": "UUID",
      "unique": false,
    },
    "updatedAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
  },
  {
    "createdAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "finished": {
      "allowNull": true,
      "defaultValue": false,
      "primaryKey": false,
      "type": "TINYINT(1)",
      "unique": false,
    },
    "metadata": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "participantId": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "references": {
        "key": "participantId",
        "model": "wwl_participants",
      },
      "type": "UUID",
      "unique": false,
    },
    "privateInfo": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "publicInfo": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "sessionId": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": true,
      "type": "UUID",
      "unique": true,
    },
    "studyId": {
      "allowNull": false,
      "defaultValue": undefined,
      "primaryKey": false,
      "references": {
        "key": "studyId",
        "model": "wwl_studies",
      },
      "type": "VARCHAR(255)",
      "unique": false,
    },
    "updatedAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
  },
  {
    "createdAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
    "deletionProtection": {
      "allowNull": false,
      "defaultValue": true,
      "primaryKey": false,
      "type": "TINYINT(1)",
      "unique": false,
    },
    "privateInfo": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "publicInfo": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "JSON",
      "unique": false,
    },
    "studyId": {
      "allowNull": false,
      "defaultValue": null,
      "primaryKey": true,
      "type": "VARCHAR(255)",
      "unique": true,
    },
    "updatedAt": {
      "allowNull": true,
      "defaultValue": undefined,
      "primaryKey": false,
      "type": "DATETIME",
      "unique": false,
    },
  },
]
`;

exports[`Database Migrations > should always create the same tables 1`] = `
[
  "wwl_deployments",
  "wwl_internal_admin_sessions",
  "wwl_internal_migrations",
  "wwl_leaderboard_scores",
  "wwl_leaderboards",
  "wwl_participants",
  "wwl_responses",
  "wwl_sessions",
  "wwl_studies",
]
`;
